Documentation and installation requirement changes

FHof opened the pull request:
To use an integrator with another numerical backend than PyTorch, for example Numpy, the only torchquad-related change needed in the example code is to pass `backend="numpy"` to the integrate method (and enabling numpy behaviour in case of Tensorflow) Other code changes are not related to torchquad, for example replacing torch.sin with np.sin in the integrand for the numpy backend or using jax.grad for gradient calculation with JAX; these operations are already well documented at other websites.

<!--
I also changed the math formula image format to SVG so that the formulas look sharp even when they are not shown in the original resolution. Here's a before and after comparison:
![latexpng](https://user-images.githubusercontent.com/23431138/152999770-2c434c35-8cee-4b07-8ffd-c28bab064165.png)
![latexsvg](https://user-images.githubusercontent.com/23431138/152999773-401eea86-6cd7-4345-b692-3f26ad57a8bb.png)
-->

gomezzz commented on torchquad/integration/vegas.py:
```diff
@@ -42,6 +42,8 @@ def integrate(
     ):
         """Integrates the passed function on the passed domain using VEGAS.
 
+        If the integrand output is far away from zero, i.e. lies within [a, a+c] for an absolutely big constant a and small constant c, VEGAS does not adapt well to the integrand. Shifting the integrand so that it is close to zero may improve the accuracy of the calculated integral in this case.
```
```suggestion
        If the integrand output is far away from zero, i.e. lies within [a, a+c] for an constant a with large absolute value and small constant c, VEGAS does not adapt well to the integrand. Shifting the integrand so that it is close to zero may improve the accuracy of the calculated integral in this case.
```
Sounds too much like some former presidents otherwise :)

FHof replied:
done.

gomezzz reviewed (changes requested):
Some comments

gomezzz commented:
As discussed:
- change docs so that user responsible is for installing backend frameworks and by default only numpy is there

FHof commented on docs/source/tutorial.rst:
```diff
@@ -456,18 +478,27 @@ We selected the Trapezoid rule and the Monte Carlo method to showcase that getti
     set_up_backend("torch", data_type="float64")
     # Number of Function evaluations
```
I assume the number was copy-pasted from [the gradient test](https://github.com/esa/torchquad/blob/d1e532e8986e5ded5ee7f371ad0071c5db1c2663/torchquad/tests/gradient_test.py#L33), where other composite Newton Cotes integrators have restrictions on the possible values for N.
I changed this number to 10000.

FHof commented on docs/source/tutorial.rst:
```diff
@@ -456,18 +478,27 @@ We selected the Trapezoid rule and the Monte Carlo method to showcase that getti
+        else:
+            result = integrator.integrate(
+                test_function, dim=1, N=N, integration_domain=domain
+            )
```
I agree. I changed the code example now.

FHof commented on docs/source/conf.py:
```diff
@@ -68,3 +68,12 @@
+
+# Use a vectorgraphic format for math formulas so that they look sharp even if a
+# browser zoom is set.
+imgmath_image_format = "svg"
```
Perhaps it doesn't work if dvisvgm is missing.
https://www.sphinx-doc.org/en/master/usage/extensions/math.html#confval-imgmath_image_format

I've removed these changes now since they are unimportant for this PR.

FHof commented:
> Can we add a sentence here how you need to install the respective backend packages you want and that by default only torch is required?

> change docs so that user responsible is for installing backend frameworks and by default only numpy is there

I've removed the torch installation from requirements.txt and similar files (which is not a documentation-only change) and
added instructions about how to install backends.
I've also mentioned the backend versions which I used all the time.

FHof commented on README.md:
````diff
@@ -99,28 +99,28 @@ For a detailed list of required packages, please refer to the [conda environment
+The easiest way to install torchquad is simply to
 
    ```sh
    conda install torchquad -c conda-forge -c pytorch
````
I've changed the files in this PR now. autoray and the backends are the only new dependencies I think.

FHof commented on README.md:
```diff
@@ -66,23 +66,23 @@
 
-* [PyTorch](https://pytorch.org/), which means it is fully differentiable and can be used for machine learning, and
-* [conda](https://docs.conda.io/en/latest/), which will take care of all requirements for you.
+* [autoray](https://github.com/jcmgray/autoray), which means the implemented quadrature supports [Numpy](https://numpy.org/) and can be used for machine learning with modules such as [PyTorch](https://pytorch.org/), [JAX](https://github.com/google/jax/) and [Tensorflow](https://www.tensorflow.org/), where it is fully differentiable
```
done.

FHof commented on docs/source/install.rst:
```diff
@@ -5,49 +5,49 @@ Getting started
 
-- `PyTorch <https://pytorch.org/>`_, which means it is fully differentiable and can be used for machine learning, and
-- `conda <https://docs.conda.io/en/latest/>`_, which will take care of all requirements for you.
+- `autoray <https://github.com/jcmgray/autoray>`_, which means the implemented quadrature supports `Numpy <https://numpy.org/>`_ and can be used for machine learning with modules such as `PyTorch <https://pytorch.org/>`_, `JAX <https://github.com/google/jax/>`_ and `Tensorflow <https://www.tensorflow.org/>`_, where it is fully differentiable
```
done.

FHof commented on docs/source/tutorial.rst:
```diff
@@ -230,8 +251,8 @@ Let’s see what type of value we get for different integrators.
     vegas = VEGAS()
-    result = vegas.integrate(f,dim=1,N=101,integration_domain=integration_domain)
-    print_error(result,solution)
+    result = vegas.integrate(f, dim=1, N=101, integration_domain=integration_domain)
```
I've tested it and it still works and always produces different values because it's non-deterministic.
The text already mentions `VEGAS currently requires a larger number of samples to function correctly (as it performs several iterations).`

When I run the code, it warns that the VEGASMap update in the warmup doesn't work, probably because there are too few samples. It also shows that it executes 60 function evaluations.
```
12:59:29|TQ-WARNING| Cannot update the VEGASMap. This can happen with an integrand which evaluates to zero everywhere.
12:59:29|TQ-WARNING| Cannot update the VEGASMap. This can happen with an integrand which evaluates to zero everywhere.
12:59:29|TQ-WARNING| Cannot update the VEGASMap. This can happen with an integrand which evaluates to zero everywhere.
12:59:29|TQ-WARNING| Cannot update the VEGASMap. This can happen with an integrand which evaluates to zero everywhere.
12:59:29|TQ-WARNING| Cannot update the VEGASMap. This can happen with an integrand which evaluates to zero everywhere.
12:59:29|TQ-INFO| Iteration 1, Acc=4.8943e-01, Result=2.1383e+01,neval=4
12:59:29|TQ-INFO| Iteration 2, Acc=5.0659e-01, Result=6.2991e+00,neval=8
12:59:29|TQ-INFO| Iteration 3, Acc=7.5145e-01, Result=1.0113e+01,neval=12
12:59:29|TQ-INFO| Iteration 4, Acc=2.2369e-01, Result=1.4703e+01,neval=16
12:59:29|TQ-INFO| Iteration 5, Acc=2.4869e-01, Result=9.2477e+00,neval=20
12:59:29|TQ-INFO| Iteration 6, Acc=2.1908e-01, Result=1.4826e+01,neval=28
12:59:29|TQ-INFO| Iteration 7, Acc=1.8378e-01, Result=1.6443e+01,neval=36
12:59:29|TQ-INFO| Iteration 8, Acc=2.5186e-01, Result=1.3002e+01,neval=44
12:59:29|TQ-INFO| Iteration 9, Acc=2.6398e-01, Result=7.4992e+00,neval=52
12:59:29|TQ-INFO| Iteration 10, Acc=1.7388e-01, Result=1.3658e+01,neval=60
12:59:29|TQ-INFO| Computed integral after 60 evals was 1.19909105e+01.
Results: 11.990910530090332
Abs. Error: 7.87201881e-01
Rel. Error: 6.16054907e-02
```
```
13:02:18|TQ-WARNING| Cannot update the VEGASMap. This can happen with an integrand which evaluates to zero everywhere.
13:02:18|TQ-WARNING| Cannot update the VEGASMap. This can happen with an integrand which evaluates to zero everywhere.
13:02:18|TQ-WARNING| Cannot update the VEGASMap. This can happen with an integrand which evaluates to zero everywhere.
13:02:18|TQ-WARNING| Cannot update the VEGASMap. This can happen with an integrand which evaluates to zero everywhere.
13:02:18|TQ-WARNING| Cannot update the VEGASMap. This can happen with an integrand which evaluates to zero everywhere.
13:02:18|TQ-INFO| Iteration 1, Acc=4.0357e-01, Result=2.2801e+01,neval=4
13:02:18|TQ-INFO| Iteration 2, Acc=6.6709e-01, Result=8.8470e+00,neval=8
13:02:18|TQ-INFO| Iteration 3, Acc=7.7849e-02, Result=1.1478e+01,neval=12
13:02:18|TQ-INFO| Iteration 4, Acc=4.5168e-01, Result=1.2777e+01,neval=16
13:02:18|TQ-INFO| Iteration 5, Acc=1.3146e-01, Result=1.8832e+01,neval=20
13:02:18|TQ-INFO| Iteration 6, Acc=3.0190e-01, Result=1.1799e+01,neval=28
13:02:18|TQ-INFO| Iteration 7, Acc=2.0750e-01, Result=8.7154e+00,neval=36
13:02:18|TQ-INFO| Iteration 8, Acc=2.3230e-01, Result=1.5856e+01,neval=44
13:02:18|TQ-INFO| Iteration 9, Acc=2.4385e-01, Result=1.3202e+01,neval=52
13:02:18|TQ-INFO| Iteration 10, Acc=3.1772e-01, Result=1.2134e+01,neval=60
13:02:18|TQ-INFO| Computed integral after 60 evals was 1.10178499e+01.
Results: 11.017849922180176
Abs. Error: 1.76026249e+00
Rel. Error: 1.37756065e-01
```

FHof commented on docs/source/tutorial.rst:
```diff
@@ -322,8 +344,8 @@ We are working on giving the user more flexibility on this point.
     vegas = VEGAS()
-    result = vegas.integrate(f_2,dim=10,N=N,integration_domain=integration_domain)
-    print_error(result,solution)
+    result = vegas.integrate(f_2, dim=10, N=N, integration_domain=integration_domain)
```
I've rerun it and the results are quite similar to before. VEGAS evaluates the function 37775 times now (N is 59049) and randomly it is more or less accurate than MonteCarlo. With N = 5**10 VEGAS is more accurate than MonteCarlo.

gomezzz commented on environment.yml:
```diff
@@ -12,12 +11,3 @@ dependencies:
 - sphinx_rtd_theme>=0.5.1
 - tqdm>=4.56.0
-# Numerical backend installations with CUDA support where possible:
-- numpy>=1.19.5
```
are you sure we don't need numpy anyways somewhere in the code?

gomezzz reviewed (approved):
minor comments
